synnax/pkg/distribution/channel/lease_proxy.go

```diff
@@ -1,301 +1,352 @@
-	if err != nil {
-		return nil, err
-	}
-	cm := CreateMessage{Channels: channels, RetrieveIfNameExists: retrieveIfNameExists}
-	res, err := lp.Transport.CreateClient().Send(ctx, addr, cm)
-	if err != nil {
-		return nil, err
-	}
-	return res.Channels, nil
-}
-
-func (lp *leaseProxy) deleteByName(ctx context.Context, tx gorp.Tx, names []string) error {
-	var res []Channel
-	if err := gorp.NewRetrieve[Key, Channel]().Entries(&res).Where(func(c *Channel) bool {
-		return lo.Contains(names, c.Name)
-	}).Exec(ctx, tx); err != nil {
-		return err
-	}
-	keys := KeysFromChannels(res)
-	return lp.delete(ctx, tx, keys)
-}
-
-func (lp *leaseProxy) delete(ctx context.Context, tx gorp.Tx, keys Keys) error {
-	batch := lp.deleteRouter.Batch(keys)
-	for nodeKey, entries := range batch.Peers {
-		err := lp.deleteRemote(ctx, nodeKey, entries)
-		if err != nil {
-			return err
-		}
-	}
-	if len(batch.Free) > 0 {
-		err := lp.deleteFreeVirtual(ctx, tx, batch.Free)
-		if err != nil {
-			return err
-		}
-	}
-	return lp.deleteGateway(ctx, tx, batch.Gateway)
-}
-
-func (lp *leaseProxy) deleteFreeVirtual(ctx context.Context, tx gorp.Tx, channels Keys) error {
-	return gorp.NewDelete[Key, Channel]().WhereKeys(channels...).Exec(ctx, tx)
-}
-
-func (lp *leaseProxy) deleteGateway(ctx context.Context, tx gorp.Tx, keys Keys) error {
-	if err := gorp.NewDelete[Key, Channel]().WhereKeys(keys...).Exec(ctx, tx); err != nil {
-		return err
-	}
-	c := errutil.NewCatch(errutil.WithAggregation())
-	for _, key := range keys {
-		c.Exec(func() error { return lp.TSChannel.DeleteChannel(key.StorageKey()) })
-	}
-	return c.Error()
-}
-
-func (lp *leaseProxy) deleteRemote(ctx context.Context, target core.NodeKey, keys Keys) error {
-	addr, err := lp.HostResolver.Resolve(target)
-	if err != nil {
-		return err
-	}
-	_, err = lp.Transport.DeleteClient().Send(ctx, addr, DeleteRequest{Keys: keys})
-	return err
-}
+// Copyright 2023 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+package channel
+
+import (
+	"context"
+
+	"github.com/samber/lo"
+	"github.com/synnaxlabs/synnax/pkg/distribution/core"
+	"github.com/synnaxlabs/synnax/pkg/distribution/ontology"
+	"github.com/synnaxlabs/synnax/pkg/distribution/ontology/group"
+	"github.com/synnaxlabs/synnax/pkg/distribution/proxy"
+	"github.com/synnaxlabs/x/errutil"
+	"github.com/synnaxlabs/x/gorp"
+	"github.com/synnaxlabs/x/kv"
+)
+
+type leaseProxy struct {
+	ServiceConfig
+	createRouter     proxy.BatchFactory[Channel]
+	deleteRouter     proxy.BatchFactory[Key]
+	leasedCounter    *kv.AtomicInt64Counter
+	freeCounter      *kv.AtomicInt64Counter
+	externalCounter  *kv.AtomicInt64Counter
+	deletedChannels  *ChannelGroup

```

Comment:
These variables should just be called `deleted` and `internal`. Again, package contextual